text/0000-memory_model.md

```diff
@@ -0,0 +1,393 @@
+- Feature Name: N/A
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Giving Rust a memory model. This allows us to understand what exactly is
+undefined behavior, and what is not, in unsafe code.
+
+# Motivation
+[motivation]: #motivation
+
+To allow unsafe code to be written without worrying about whether the compiler
+will miscompile your code.  Our current system is ill-defined, and far too
+cautious (and in other cases, completely undefined; for example, what are the
+semantics of raw pointer aliasing?).
+
+# Detailed design
+[design]: #detailed-design
+
+This is the complicated part :)
+
+## Using a value
+
+Anything which touches a value, and is not either a move or copy from the value,
+move or copy into an lvalue, or taking a reference to the value, is a use of the
+value. Examples of uses are: arithmetic, match, indexing. Examples of things
+which are not uses are: returning, passing a value to a function, `let x = y`,
+and `let x = &y`. A reborrow is not a use.
+
+## Type Representation
+
+Each type `T` shall be equivalent to a byte array: `[u8; size_of::<T>()]`. Each
```

Comment:
Before making statements like this it would be useful to give some motivation, discuss [alternatives](https://github.com/rust-lang/rfcs/issues/1397), their pros and cons, and why we need to abandon them forever in favor of this approach.
This text contains a lot of statements.